apps/hadash/hadash.app.js

```diff
@@ -0,0 +1,227 @@
+    let HAresp = JSON.parse(data.resp);
+    let title4prompt = title;
+    let msg = HAresp.state;
+    if ('attributes' in HAresp) {
+      if ('friendly_name' in HAresp.attributes)
+        title4prompt = HAresp.attributes.friendly_name;
+      if ('unit_of_measurement' in HAresp.attributes)
+        msg += HAresp.attributes.unit_of_measurement;
+    }
+    E.showPrompt(msg, { title: title4prompt, buttons: {OK: true} }).then((v) => { E.showMenu(menus[level]); });
+  }).catch( error => {
+    console.log(error);
+    E.showPrompt('Error querying state!', { title: title, buttons: {OK: true} }).then((v) => { E.showMenu(menus[level]); });
+  });
+}
+
+
+// call a service
+function callService(title, domain, service, data, level) {
+  E.showMessage('Calling HA service', { title: title });
+  Bangle.http(settings.HAbaseUrl+'/services/'+domain+'/'+service, {
+    method: 'POST',
+    body: data,
+    headers: {
+      'Authorization': 'Bearer '+settings.HAtoken,
+      'Content-Type': 'application/json'
+    },
+  }).then(data => {
+    //console.log(data);
+    E.showPrompt('Service called successfully', { title: title, buttons: {OK: true} }).then((v) => { E.showMenu(menus[level]); });
+  }).catch( error => {
+    console.log(error);
+    E.showPrompt('Error calling service!', { title: title, buttons: {OK: true} }).then((v) => { E.showMenu(menus[level]); });
+  });
+}
+
+
+// callbacks for service input menu entries
+function serviceInputChoiceChange(v, key, entry, level) {
+  entry.input[key].value = entry.input[key].options[v];
+  getServiceInputData(entry, level);
+}
+
+function serviceInputFreeform(key, entry, level) {
+  require("textinput").input({text: entry.input[key].value}).then(result => {
+    entry.input[key].value = result;
+    getServiceInputData(entry, level);
+  });
+}
+
+// get input data before calling a service
+function getServiceInputData(entry, level) {
+  let serviceInputMenu = {
+    '': {
+      'title': entry.title,
+      'back': () => E.showMenu(menus[level])
+    },
+  };
+  let CBs = {};
+  for (let key in entry.input) {
+    // pre-fill data with default values
+    if ('value' in entry.input[key])
+      entry.data[key] = entry.input[key].value;
+
+    let label = ( ('label' in entry.input[key] && entry.input[key].label) ? entry.input[key].label : key );
+    let key4CB = key;
+
+    if ('options' in entry.input[key] && entry.input[key].options.length) {
+      // give choice from a selection of options
+      let idx = -1;
+      for (let i in entry.input[key].options) {
+        if (entry.input[key].value == entry.input[key].options[i]) {
+          idx = i;
+        }
+      }
+      if (idx == -1) {
+        idx = entry.input[key].options.push(entry.input[key].value) - 1;
+      }
+      // the setTimeout method can not be used for the "format" CB since it expects a return value - using eval instead:
+      eval('CBs["'+key+'_format"] = function(v) { return entry.input["'+key+'"].options[v]; }');
```

Comment:
Can this be:

```suggestion
      CBs[`${key}_format`] = function(v) { return entry.input[key].options[v]; }
```

Reply:
I guess the ```CBs[`${key}_format`]``` part would work, but the anonymous function won't have the "key" variable available. So, we still need eval for that.

Reply:
Ah gotya! For that we can do something like:
```suggestion
      CBs[`${key}_format`] = ((key) => function(v) { return entry.input[key].options[v]; })(key);
```